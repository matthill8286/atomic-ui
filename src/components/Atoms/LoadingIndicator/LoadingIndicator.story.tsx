import { number, radios } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { LoadingIndicator, LoadingIndicatorColor, LoadingIndicatorProps } from './LoadingIndicator'

const loadingIndicatorColor: { [key: string]: LoadingIndicatorColor } = {
  white: 'white',
  black: 'black',
  primary: 'primary',
}

const knobs = (): LoadingIndicatorProps => {
  return {
    isVisible: true,
    color: radios('Color', loadingIndicatorColor, loadingIndicatorColor.primary),
    size: number('Size', 50),
  }
}

storiesOf('Design System/Atoms/LoadingIndicator', module).add('Default', () => (
  <LoadingIndicator {...knobs()} />
))
