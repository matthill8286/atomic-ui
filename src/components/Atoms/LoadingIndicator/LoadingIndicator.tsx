import React from 'react'
import { rotateForward } from '@/styles/animation'
import { styled } from '@/styles/styled'

export type LoadingIndicatorColor = 'white' | 'black' | 'primary'

export interface StyledLoadingIndicatorProps {
  isVisible: boolean
  size: number
  barWidth: number
  color: LoadingIndicatorColor
}

export interface LoadingIndicatorProps {
  isVisible: boolean
  size?: number
  barWidth?: number
  color?: LoadingIndicatorColor
}

const StyledLoadingWrapper = styled.div<StyledLoadingIndicatorProps>`
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  display: ${({ isVisible }) => (isVisible ? 'block' : 'none')};
  position: absolute;

  & div {
    box-sizing: border-box;
    display: block;
    position: absolute;
    width: 100%;
    height: 100%;
    border: ${({ size }) => size * 0.1}px solid;
    border-radius: 50%;
    animation: ${rotateForward} 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
    border-color: ${({ theme, color }) => theme.color[color]} transparent transparent transparent;
  }

  & div:nth-child(1) {
    animation-delay: -0.45s;
  }

  & div:nth-child(2) {
    animation-delay: -0.3s;
  }

  & div:nth-child(3) {
    animation-delay: -0.15s;
  }
`

export const StyledLoadingWrapperAlternate = styled.ul<StyledLoadingIndicatorProps>`
  position: absolute;
  display: ${({ isVisible }) => (isVisible ? 'flex' : 'none')};
  margin: 0 auto;
  padding: 0;
  justify-content: center;
  height: ${({ size }) => size * 0.3}px;
  li {
    list-style: none;
    width: ${({ barWidth }) => barWidth * 0.05}px;
    height: ${({ size }) => size * 0.3}px;
    margin: 0 3px;
    animation: animate 0.4s infinite alternate;
    background: ${({ theme, color }) => theme.color[color]};

    &:nth-child(2) {
      animation-delay: 0.1s;
    }
    &:nth-child(3) {
      animation-delay: 0.2s;
    }
  }

  @keyframes animate {
    0% {
      transform: scaleY(1);
    }
    25% {
      transform: scaleY(1);
    }
    50% {
      transform: scaleY(1);
    }
    75% {
      transform: scaleY(1);
    }
    100% {
      transform: scaleY(1.7);
    }
  }
`

export const LoadingIndicator: React.FC<LoadingIndicatorProps> = ({
  isVisible,
  size = 50,
  barWidth = 15,
  color = 'primary',
  ...otherProps
}) => {
  return (
    <StyledLoadingWrapper
      isVisible={isVisible}
      size={size}
      barWidth={barWidth}
      color={color}
      {...otherProps}>
      <div />
      <div />
      <div />
      <div />
    </StyledLoadingWrapper>
  )
}
