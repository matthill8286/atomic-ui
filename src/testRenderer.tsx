import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
