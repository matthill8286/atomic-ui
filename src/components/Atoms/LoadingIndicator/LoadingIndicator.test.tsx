import * as React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { LoadingIndicator } from './LoadingIndicator'

describe('LoadingIndicator', () => {
  it('renders filtered theme', () => {
    const tree = renderWithTheme(<LoadingIndicator isVisible={false} />)
    expect(tree).toMatchSnapshot()
  })
})
