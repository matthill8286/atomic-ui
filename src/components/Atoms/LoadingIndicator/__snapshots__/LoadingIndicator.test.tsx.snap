// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator renders filtered theme 1`] = `
.c0 {
  width: 50px;
  height: 50px;
  display: none;
  position: absolute;
}

.c0 div {
  box-sizing: border-box;
  display: block;
  position: absolute;
  width: 100%;
  height: 100%;
  border: 5px solid;
  border-radius: 50%;
  -webkit-animation: fEWCgj 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  animation: fEWCgj 1.2s cubic-bezier(0.5,0,0.5,1) infinite;
  border-color: #00C8DC transparent transparent transparent;
}

.c0 div:nth-child(1) {
  -webkit-animation-delay: -0.45s;
  animation-delay: -0.45s;
}

.c0 div:nth-child(2) {
  -webkit-animation-delay: -0.3s;
  animation-delay: -0.3s;
}

.c0 div:nth-child(3) {
  -webkit-animation-delay: -0.15s;
  animation-delay: -0.15s;
}

<div
  className="c0"
  color="primary"
  size={50}
>
  <div />
  <div />
  <div />
  <div />
</div>
`;
